notebooks/02_model/hypertune_spark_deep_dive.ipynb

```diff
@@ -0,0 +1,983 @@
+       "    </tr>\n",
+       "  </thead>\n",
+       "  <tbody>\n",
+       "    <tr>\n",
+       "      <th>0</th>\n",
+       "      <td>spark</td>\n",
+       "      <td>0.993568</td>\n",
+       "      <td>0.787028</td>\n",
+       "      <td>0.244673</td>\n",
+       "      <td>0.208661</td>\n",
+       "      <td>44.237093</td>\n",
+       "    </tr>\n",
+       "    <tr>\n",
+       "      <th>0</th>\n",
+       "      <td>hyperopt</td>\n",
+       "      <td>0.968426</td>\n",
+       "      <td>0.774891</td>\n",
+       "      <td>0.295251</td>\n",
+       "      <td>0.248203</td>\n",
+       "      <td>116.420096</td>\n",
+       "    </tr>\n",
+       "    <tr>\n",
+       "      <th>0</th>\n",
+       "      <td>sample</td>\n",
+       "      <td>0.985228</td>\n",
+       "      <td>0.780406</td>\n",
+       "      <td>0.258821</td>\n",
+       "      <td>0.221890</td>\n",
+       "      <td>211.739382</td>\n",
+       "    </tr>\n",
+       "  </tbody>\n",
+       "</table>\n",
+       "</div>"
+      ],
+      "text/plain": [
+       "   Approach      RMSE       MAE  Explained variance  R squared     Elapsed\n",
+       "0     spark  0.993568  0.787028            0.244673   0.208661   44.237093\n",
+       "0  hyperopt  0.968426  0.774891            0.295251   0.248203  116.420096\n",
+       "0    sample  0.985228  0.780406            0.258821   0.221890  211.739382"
+      ]
+     },
+     "execution_count": 94,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "comparison"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "From the results, it can be seen that, *with the same number of iterations*, Spark native construct based approach takes the least amount of time, and not suprisingly, sample based approach takes the most amount of time. Intuitively this is because Spark native constructs leverage the underlying Java codes for running the actual analytics with high performance efficiency. \n",
+    "\n",
+    "The three approaches use the same RMSE loss. In this measure, the `hyperopt` based approach performs the best. This may be owing to the TPE searching algorithm used in `hyperopt`. Note the differences in the RMSE metrics may also come from the randomness of the intermediate steps in parameter tuning process. In practice, multiple runs are required for generating statistically robust comparison results. We have tried 5 times for running the same comparison codes above. The results aligned well with each other in terms of objective metric values and elapsed time. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Conclusions"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In summary, there are mainly three different approaches for running hyperparameter tuning for Spark based recommendation algorithm. \"How-to\" guides and Pros and Cons are tabulated below."
```

Comment:
I really like this format for conclusions, it states pros and cons of each method. Maybe we should follow this pattern in other notebooks?

Reply:
Good idea. If there is no comparative study, a general conclusion is great to have as key takeaways of the notebook.